File: tests/features/steps/demo-steps.py
from behave import *
from prelude_sdk.models.account import Account
from prelude_sdk.controllers.detect_controller import DetectController

@given('I am signed in as a demo user')
def step_impl(context):
    # No-op:  Tests are always available to an account
    context.controller = DetectController(account=Account())

# --- Activity data scenarios ---

@given('I want to request activity data from the "{view}" view')
def step_impl(context, view):
    context.controller = DetectController(account=Account())
    context.view = view
    context.options = {}

@given('I specify a test ID "{test}"')
def step_impl(context, test):
    context.options["tests"] = [test]

@given('I specify a DOS "{dos}"')
def step_impl(context, dos):
    context.options["dos"] = [dos]

@given('I specify a tag "{tag}"')
def step_impl(context, tag):
    context.options["tags"] = [tag]

@when('I retrieve the activity data')
def step_impl(context):
    context.response = context.controller.describe_activity(filters=context.options, view=context.view)

@then('the response should contain logs')
def step_impl(context):
    assert isinstance(context.response, list)
    assert len(context.response) > 0
    assert "id" in context.response[0]
    assert "date" in context.response[0]
    assert "test" in context.response[0]
    assert "endpoint_id" in context.response[0]
    assert "status" in context.response[0]
    assert "dos" in context.response[0]
    assert "tags" in context.response[0]
    assert "edr_id" in context.response[0]

@then('the response should contain data only for that test ID')
def step_impl(context):
    assert all(ele["test"] == context.options["tests"][0] for ele in context.response)

@then('the response should contain data only for that tag')
def step_impl(context):
    assert all(ele["tags"][0] == context.options["tags"][0] for ele in context.response)

@then('the response should contain data only for that DOS')
def step_impl(context):
    assert all(ele["dos"] == context.options["dos"][0] for ele in context.response)

@then('the response should contain insights')
def step_impl(context):
    assert isinstance(context.response, list)
    assert len(context.response) > 0
    assert "dos" in context.response[0]
    assert "test" in context.response[0]
    assert "volume" in context.response[0]
    assert "error" in context.response[0]['volume']
    assert "protected" in context.response[0]['volume']
    assert "unprotected" in context.response[0]['volume']


@then('the response should contain probes')
def step_impl(context):
    assert isinstance(context.response, list)
    assert len(context.response) > 0
    assert "endpoint_id" in context.response[0]
    assert "dos" in context.response[0]
    assert "tags" in context.response[0]
    assert "state" in context.response[0]
    assert "edr_id" in context.response[0]

@then('the response should contain rules')
def step_impl(context):
    assert isinstance(context.response, list)
    assert len(context.response) > 0
    assert "rule" in context.response[0]
    assert "id" in context.response[0]['rule']
    assert "label" in context.response[0]['rule']
    assert "published" in context.response[0]['rule']
    assert "description" in context.response[0]['rule']
    assert "long_description" in context.response[0]['rule']

#  --- Scenario: Get tests ---

@when('I retrieve tests')
def step_impl(context):
    context.response = context.controller.list_tests()

@then('the response should contain tests')
def step_impl(context):
    assert isinstance(context.response, list)
    assert len(context.response) > 0
    assert "id" in context.response[0]
    assert "name" in context.response[0]
    assert "account_id" in context.response[0]

# --- Scenario: Search the NVD for a keyword ---

@given('I have a keyword to search for "{text}"')
def step_impl(context, text):
    context.keyword = text

@when('I search for the keyword')
def step_impl(context):
    context.response = context.controller.search(identifier=context.keyword)

@then('the response should contain search results')
def step_impl(context):
    assert 'info' in context.response
    assert 'tests' in context.response

File: tests/features/javascript/demo-steps.mjs
import { Given, Then, When } from "@cucumber/cucumber";
import { Service } from "../../../javascript/sdk/dist/sdk.js";
import assert from "assert";

Given("I am signed in as a demo user", function () {
  this.controller = new Service({
    host: "https://api.staging.preludesecurity.com",
    credentials: {
      account: "prelude-demo",
      token: "executive-demo",
    },
  }).detect;
});

/* --- Activity data scenarios --- */

Given(
  "I want to request activity data from the {string} view",
  function (view) {
    this.view = view;
    this.options = {};
  }
);

Given("I specify a test ID {string}", function (test) {
  this.options.tests = [test];
});

Given("I specify a DOS {string}", function (dos) {
  this.options.dos = [dos];
});

Given("I specify a tag {string}", function (tag) {
  this.options.tags = [tag];
});

When("I retrieve the activity data", async function () {
  this.response = await this.controller.describeActivity({
    ...this.options,
    view: this.view,
  });
});

Then("the response should contain logs", async function () {
  assert(Array.isArray(this.response));
  assert(this.response.length > 0);
  assert("id" in this.response[0], "no id in response");
  assert("date" in this.response[0], "no date in response");
  assert("test" in this.response[0], "no test in response");
  assert("endpoint_id" in this.response[0], "no endpoint_id in response");
  assert("status" in this.response[0], "no status in response");
  assert("dos" in this.response[0], "no dos in response");
  assert("tags" in this.response[0], "no tags in response");
  assert("edr_id" in this.response[0], "no edr_id in response");
});

Then("the response should contain data only for that test ID", function () {
  assert(
    this.response.every(
      (data) => data.test === this.options.tests[0],
      "response contains tests not specified"
    )
  );
});

Then("the response should contain data only for that tag", function () {
  assert(
    this.response.every(
      (data) => data.tags[0] === this.options.tags[0],
      "response contains tag not specified"
    )
  );
});

Then("the response should contain data only for that DOS", function () {
  assert(
    this.response.every(
      (data) => data.dos === this.options.dos[0],
      "response contains DOS not specified"
    )
  );
});

Then("the response should contain insights", function () {
  assert(Array.isArray(this.response));
  assert(this.response.length > 0);
  assert("dos" in this.response[0], "no dos in response");
  assert("test" in this.response[0], "no test in response");
  assert("volume" in this.response[0], "no volume in response");
  assert("error" in this.response[0].volume, "no error in volume response");
  assert(
    "protected" in this.response[0].volume,
    "no protected in volume response"
  );
  assert(
    "unprotected" in this.response[0].volume,
    "no unprotected in volume response"
  );
});

Then("the response should contain probes", function () {
  assert(Array.isArray(this.response));
  assert(this.response.length > 0);
  assert("endpoint_id" in this.response[0], "no endpoint_id in response");
  assert("dos" in this.response[0], "no dos in response");
  assert("tags" in this.response[0], "no tags in response");
  assert("state" in this.response[0], "no state in response");
  assert("edr_id" in this.response[0], "no edr_id in response");
});

Then("the response should contain rules", function () {
  assert(Array.isArray(this.response));
  assert(this.response.length > 0);
  assert("rule" in this.response[0], "no rule in response");
  assert("id" in this.response[0].rule, "no id in rule response");
  assert("label" in this.response[0].rule, "no label in rule response");
  assert("published" in this.response[0].rule, "no published in rule response");
  assert(
    "description" in this.response[0].rule,
    "no description in rule response"
  );
  assert(
    "long_description" in this.response[0].rule,
    "no long_description in rule response"
  );
});

/* --- Scenario: Get tests --- */

When("I retrieve tests", async function () {
  this.response = await this.controller.listTests();
});

Then("the response should contain tests", function () {
  assert(Array.isArray(this.response), "response is not an array");
  assert(this.response.length > 0, "response is empty");
  assert("id" in this.response[0], "id not in response");
  assert("name" in this.response[0], "name not in response");
  assert("account_id" in this.response[0], "account not in response");
});

/* --- Scenario: Search the NVD for a keyword --- */

Given("I have a keyword to search for {string}", function (keyword) {
  this.keyword = keyword;
});

When("I search for the keyword", async function () {
  this.response = await this.controller.search(this.keyword);
});

Then("the response should contain search results", function () {
  assert("info" in this.response);
  assert("tests" in this.response);
});
